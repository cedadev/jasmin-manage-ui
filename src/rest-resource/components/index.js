// Export the delete button and status components directly
export * from './InstanceDeleteButton';
export * from './Status';

// Export the form pieces in their own namespace
import * as Form from './ResourceForm';
export { Form };
