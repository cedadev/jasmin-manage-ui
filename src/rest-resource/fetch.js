import Cookies from 'js-cookie';


class HttpError extends Error {
    constructor(status, statusText, responseText) {
        super(statusText);
        this.name = this.constructor.name;
        this.status = status;
        this.statusText = statusText;
        this.responseText = responseText;
    }

    json() {
        return JSON.parse(this.responseText)
    }
}


// Utility function that asynchronously waits for the specified duration
const sleep = (ms) => (new Promise(resolve => setTimeout(resolve, ms)));


const apiFetch = async (url, {
    method = 'GET',
    headers = {},
    data,
    body,
    ...options
} = {}) => {
    // Populate the required headers for the request
    const defaultHeaders = {};
    // For POST/PUT/DELETE, declare the content type and include the CSRF token if present
    if( ['POST', 'PUT', 'DELETE'].includes(method.toUpperCase()) ) {
        defaultHeaders['Content-Type'] = 'application/json';
        const csrfToken = Cookies.get('csrftoken');
        if( csrfToken ) defaultHeaders['X-CSRFToken'] = csrfToken;
    }
    // Make the actual request, injecting the cookie credentials and headers
    const response = await fetch(url, {
        method,
        headers: { ...defaultHeaders, ...headers },
        // If data is given, encode it as JSON and use it as the request body
        // If not, use the given body
        body: data ? JSON.stringify(data) : body,
        // Always include credentials
        credentials: 'include',
        // Forward any other options that were supplied
        ...options
    });
    // For a 204, just return
    if( response.status === 204 ) return;
    // Any other successful response should be JSON
    if( response.ok ) return await response.json();
    // An error response may not be JSON, so read the response as text
    const responseText = await response.text();
    throw new HttpError(response.status, response.statusText, responseText);
};


// Export method-specific versions of apiFetch
export default apiFetch;
