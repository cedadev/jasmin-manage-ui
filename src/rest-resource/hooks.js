import { useCallback, useEffect, useState } from 'react';

import apiFetch from './fetch';


/**
 * Hook that fetches data from an HTTP endpoint and passes it to a callback.
 */
export const useOnFetch = (url, { onFetch, autoFetch = true } = {}) => {
    // Indicates whether a fetch is currently in progress
    // Setting this to true initially will cause the data to be loaded
    const [fetching, setFetching] = useState(autoFetch);
    // The error from the last fetch if it was not successful
    const [fetchError, setFetchError] = useState(null);

    // When a fetch is triggered by changing the state, fetch the data
    // Doing it this way ensures that we never have two fetches competing, as the fetch
    // is only triggered when fetching changes from false to true
    useEffect(
        () => {
            if( !fetching ) return;
            // useEffect expects the return value to be a cleanup function
            // This means we can't pass an async function directly as that returns a promise
            // Hence we have to make and call an anonymous function
            (async () => {
                try {
                    // Fetch the data from the endpoint
                    const result = await apiFetch(url);
                    // Update the data
                    onFetch(result);
                    // Clear any stored error
                    setFetchError(null);
                }
                catch(error) {
                    // For an error response, store the response
                    setFetchError(error);
                }
                finally {
                    // In all cases, the fetch is over
                    setFetching(false);
                }
            })();
        },
        [fetching]
    );

    // This function triggers a fetch by updating the state
    const fetch = useCallback(() => setFetching(true), []);

    // Return the public state for consumers to use
    return { fetching, fetchError, fetch };
};


/**
 * Hook for fetching data from an HTTP endpoint.
 */
export const useEndpoint = (url, { initialData, autoFetch = true } = {}) => {
    // Indicates if the endpoint has been successfully fetched at least once
    const [initialised, setInitialised] = useState(initialData !== undefined);
    // Make a variable to store the data and use it to initialise the hook
    const [data, setData] = useState(initialData);
    // When there is a successful fetch, set data and initialised
    const onFetch = data => { setData(data); setInitialised(true); };
    // Return the public state for consumers to use
    return { ...useOnFetch(url, { onFetch, autoFetch }), initialised, data };
};


/**
 * Hook for running a mutation against an HTTP endpoint.
 */
export const useMutation = (url, method) => {
    const [inProgress, setInProgress] = useState(false);
    // This function runs the mutation
    const mutate = async data => {
        setInProgress(true);
        try {
            // Attempt the mutation
            return await apiFetch(url, { method, data });
        }
        finally {
            // In all cases, the mutation is over
            setInProgress(false);
        }
    };
    // Return the public state for consumers to use
    return { inProgress, mutate };
};


/**
 * Hook for using an endpoint as a REST resource.
 */
export const useResource = (resourceUrl, { initialData, autoFetch = true } = {}) => {
    // Indicates if the resource has been successfully fetched at least once
    const [initialised, setInitialised] = useState(initialData !== undefined);
    // We store the current data as an object mapping id => instance
    const [data, setData] = useState(initialData || {});

    // This function updates the data of a single instance in the mapping
    const setInstanceData = id => transform => setData(state => {
        const nextInstanceState = transform instanceof Function ?
            transform(state[id]) :
            transform;
        // Allow the instance to set state to undefined to remove itself
        if( nextInstanceState !== undefined ) {
            return { ...state, [id]: nextInstanceState };
        }
        else {
            const { [id]: _, ...nextState } = state;
            return nextState;
        }
    });

    // When there is a successful fetch, set data and initialised
    // The data will be received in list form, which we transform into the object format
    const onFetch = instanceList => {
        setData(
            Object.assign(
                {},
                ...instanceList.map(instanceData => ({
                    [instanceData.id]: newInstance(
                        instanceData,
                        setInstanceData(instanceData.id)
                    )
                }))
            )
        )
        setInitialised(true);
    };

    // Initialise the on-fetch hook that will be responsible for fetching
    // the resource list
    const fetchState = useOnFetch(resourceUrl, { onFetch, autoFetch });

    // Use a mutation to create new instances and add them to the data
    const { inProgress: creating, mutate } = useMutation(resourceUrl, 'POST');
    const create = async data => {
        const instanceData = await mutate(data);
        setData(state => ({
            ...state,
            [instanceData.id]: newInstance(
                instanceData,
                setInstanceData(instanceData.id)
            )
        }));
    };

    // Return the pieces that comprise the public interface
    return { ...fetchState, data, initialised, creating, create };
};


/**
 * Hook for using a single resource instance independently of the resource.
 */
export const useResourceInstance = (instanceUrl, { initialData, autoFetch = true } = {}) => {
    // Indicates if the instance has been successfully fetched at least once
    const [initialised, setInitialised] = useState(initialData !== undefined);
    // We store the current data as an object mapping id => instance
    const [data, setData] = useState(initialData);
    // When there is a successful fetch, set data and initialised
    const onFetch = data => { setData(data); setInitialised(true); };
    // Initialise the fetch state
    const fetchState = useOnFetch(instanceUrl, { onFetch, autoFetch });

    // Use mutations to update and delete the resource, updating the data as required
    const { inProgress: updating, mutateUpdate } = useMutation(instanceUrl, 'PUT');
    const update = async data => setData(await mutateUpdate(data));

    const { inProgress: deleting, mutateDelete } = useMutation(instanceUrl, 'DELETE');
    const delete_ = async () => setData(await mutateDelete());

    // Executing an action is slightly more difficult as the URL depends on the action
    const [actionInProgress, setActionInProgress] = useState(false);
    const executeAction = async (name, actionData) => {
        setActionInProgress(true);
        try {
            // Attempt to apply the action
            const instanceData = await apiFetch(
                data._links[name],
                { method: 'POST', data: actionData }
            );
            setData(instanceData);
        }
        finally {
            setActionInProgress(false);
        }
    };

    // Return the pieces that comprise the public interface
    return {
        ...fetchState,
        data,
        initialised,
        updating: updating || actionInProgress,
        update,
        deleting,
        delete: delete_,
        executeAction
    };
};


// The setState function received by newInstance is scoped to update the data for the resource
const newInstance = (initialData, setState) => ({
    // Indicates if the instance is currently being fetched
    fetching: false,
    // Indicates if the instance is currently being updated
    updating: false,
    // Indicates if the instance is currently being deleted
    deleting: false,
    // The error from the last fetch if it was not successful
    fetchError: null,
    // The current data for the instance
    data: initialData,
    // Re-fetch the data for the instance
    fetch: async () => {
        // Mark the fetch as in progress
        setState(state => ({ ...state, fetching: true }));
        try {
            // Attempt to fetch the instance
            const instanceData = await apiFetch(initialData._links.self);
            // Update the state if successful
            setState(state => ({ ...state, fetching: false, data: instanceData }));
        }
        catch(error) {
            // If the fetch is not successful, store the error
            setState(state => ({ ...state, fetching: false, fetchError: error }));
        }
    },
    // Update the instance with new data
    update: async data => {
        // Make the update as in progress
        setState(state => ({ ...state, updating: true }));
        try {
            // Attempt to update the instance
            const instanceData = await apiFetch(initialData._links.self, { method: 'PUT', data });
            // Update the state if the update was successful
            setState(state => ({ ...state, updating: false, data: instanceData }));
        }
        catch(error) {
            // On failure, mark the update as over before re-throwing the error
            setState(state => ({ ...state, updating: false }));
            throw error;
        }
    },
    // Delete the instance
    delete: async () => {
        // Mark the delete as in progress
        setState(state => ({ ...state, deleting: true }));
        try {
            // Attempt to delete the instance
            await apiFetch(initialData._links.self, { method: 'DELETE' });
            // Remove the entry if the delete is successful
            setState(undefined);
        }
        catch(error) {
            // On failure, mark the delete as over before re-throwing the error
            setState(state => ({ ...state, deleting: false }));
            throw error;
        }
    },
    // Executes a named action on the instance
    executeAction: async (name, data) => {
        // Mark the instance as updating
        setState(state => ({ ...state, updating: true }));
        try {
            // Attempt to apply the action
            const instanceData = await apiFetch(initialData._links[name], { method: 'POST', data });
            // Update the state if the action was successful
            setState(state => ({ ...state, updating: false, data: instanceData }));
        }
        catch(error) {
            // On failure, mark the update as over before re-throwing the error
            setState(state => ({ ...state, updating: false }));
            throw error;
        }
    }
});
