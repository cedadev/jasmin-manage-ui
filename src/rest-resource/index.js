export * from './components';
export * from './context';
export * from './hooks';

import apiFetch from './fetch';
export { apiFetch };
