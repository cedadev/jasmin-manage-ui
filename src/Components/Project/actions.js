import { useNestedResource, useAggregateResource } from '../../rest-resource';

import { useConsortia, useCurrentUser } from '../../api';


/**
 * Hook to access the current user's roles for the given project.
 */
export const useProjectUserRoles = project => {
    const currentUser = useCurrentUser();
    // Extract the consortium for the project
    // If the consortia are not loaded, return nothing
    const consortia = useConsortia();
    const consortium = consortia.data[project.data.consortium];
    // This is not a fetch point for collaborators - we just want to use the
    // information if it has been loaded elsewhere
    const collaborators = useNestedResource(project, "collaborators", { fetchPoint: false });

    // Get the user's role, either from the collaborator record (if one exists) or from the project
    const userCollaborator = Object.values(collaborators.data)
        .filter(c => c.data.user.id === currentUser.data.id)
        .shift();
    const userRole = userCollaborator ?
        userCollaborator.data.role :
        project.data.current_user_role;

    return {
        // If the user has a role, they are a collaborator
        isProjectCollaborator: !!userRole,
        // If the user role is OWNER, then they are an owner
        isProjectOwner: userRole && userRole === "OWNER",
        // For this, we need to check the consortium
        isConsortiumManager: (
            // Wait for the consortia to load before deciding
            consortium &&
            consortium.data.manager.id === currentUser.data.id
        )
    };
};


/**
 * Hook to access the current user's permissions for the given project.
 */
export const useProjectPermissions = project => {
    const projectRoles = useProjectUserRoles(project);
    return {
        // The project can be edited by collaborators when in editable mode
        canEditRequirements: (
            project.data.status === "EDITABLE" &&
            projectRoles.isProjectCollaborator
        ),
        // The project can be submitted for review by a project owner
        // when in editable mode
        canSubmitForReview: (
            project.data.status === "EDITABLE" &&
            projectRoles.isProjectOwner
        ),
        // Requirements can be approved by consortium managers, but only when
        // the project is under review
        canApproveRequirements: (
            project.data.status === "UNDER_REVIEW" &&
            projectRoles.isConsortiumManager
        ),
        // Changes can be requested by the consortium manager for a project
        // that is under review
        canRequestChanges: (
            project.data.status === "UNDER_REVIEW" &&
            projectRoles.isConsortiumManager
        ),
        // A project that is under review can be submitted for provisioning by
        // the consortium manager
        canSubmitForProvisioning: (
            project.data.status === "UNDER_REVIEW" &&
            projectRoles.isConsortiumManager
        )
    };
};


/**
 * Hook to access the available actions for the current project.
 */
export const useProjectActions = project => {
    // Get all the project requirements by aggregating across the services
    // However this is not a fetch point for services - we only use them once
    // they have been loaded elsewhere in the component tree
    const services = useNestedResource(project, "services", { fetchPoint: false });
    const requirements = useAggregateResource(services, "requirements");

    // Work out if the project has requirements with various statuses
    const hasRequirementsWithStatus = status => {
        const count = Object.values(requirements.data)
            .filter(requirement => requirement.data.status === status)
            .length;
        return count > 0;
    };
    const hasRequestedRequirements = hasRequirementsWithStatus("REQUESTED");
    const hasRejectedRequirements = hasRequirementsWithStatus("REJECTED");
    const hasApprovedRequirements = hasRequirementsWithStatus("APPROVED");

    // All these actions are not allowed while the project is updating
    return {
        // Requirements cannot be edited while the project is updating
        allowEditRequirements: !project.updating,
        // The project can only be submitted for review when there are no
        // rejected requirements and at least one requirement in either the
        // requested or approved state
        // That is to cover for the case where a requirement was rejected
        // alongside others that were approved, and then subsequently removed
        allowSubmitForReview: (
            !project.updating &&
            !hasRejectedRequirements &&
            (hasRequestedRequirements || hasApprovedRequirements)
        ),
        // Requirements cannot be approved while the project is updating
        allowApproveRequirements: !project.updating,
        // Changes can be requested for a project only when:
        //   1. All requested requirements have been resolved
        //   2. There is at least one rejected requirement
        allowRequestChanges: (
            !project.updating &&
            !hasRequestedRequirements &&
            hasRejectedRequirements
        ),
        // A project can be submitted for provisioning only once:
        //   1. All the requested requirements have been resolved
        //   2. There are no rejected requirements
        //   3. There is at least one approved requirement
        allowSubmitForProvisioning: (
            !project.updating &&
            !hasRequestedRequirements &&
            !hasRejectedRequirements &&
            hasApprovedRequirements
        )
    };
};
