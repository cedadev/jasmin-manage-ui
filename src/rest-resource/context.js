import React, { createContext, useContext, useEffect } from 'react';

import { useEndpoint, useResource, useInstance } from './hooks';


// This higher order function turns a hook into a context, allowing the hook
// state to be shared between components
const createContextForHook = hookFunction => (url, options = {}) => {
    // Make a new context
    const Context = createContext();
    // Return a provider and hook that use the context to share the hook state
    return {
        // The provider is an anchor for the data and also a fetch point
        // However, we do not auto-fetch the data as soon as the provider as mounted
        Provider: ({ children }) => {
            const fetchable = hookFunction(url, { ...options, autoFetch: false });
            return <Context.Provider value={fetchable}>{children}</Context.Provider>;
        },
        // The first time a component using the hook is mounted, the data for the fetchable
        // is initialised if it has not already been initialised
        hook: (autoFetch = true) => {
            const fetchable = useContext(Context);
            useEffect(
                // If the fetchable is not initialised, trigger a fetch by marking it as dirty
                // This should not trigger multiple fetches
                () => { if( autoFetch && !fetchable.initialised ) fetchable.markDirty(); },
                []
            );
            return fetchable;
        }
    };
};


export const createContextForEndpoint = createContextForHook(useEndpoint);
export const createContextForResource = createContextForHook(useResource);
export const createContextForInstance = createContextForHook(useInstance);
