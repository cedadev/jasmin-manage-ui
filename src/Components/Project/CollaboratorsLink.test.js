import React from "react";
import { render, screen, fireEvent, waitFor } from "@testing-library/react";
import {
  CollaboratorsModalBody,
  ProjectCollaboratorsLink,
} from "./CollaboratorsLink";
import { useNestedResource } from "../../rest-resource";
import { useProjectPermissions } from "./actions";

// Mock all dependencies
jest.mock("../../rest-resource", () => ({
  Status: {
    Many: ({ children }) => children,
    Loading: ({ children }) => children,
    Unavailable: ({ children }) => children,
    Available: ({ children }) => children,
  },
  useNestedResource: jest.fn(),
}));

jest.mock("./actions", () => ({
  useProjectPermissions: jest.fn(),
}));

jest.mock("./CollaboratorsListItems", () => ({
  CollaboratorsListItems: () => <div data-testid="collaborators-list" />,
}));

jest.mock("./InvitationsListItems", () => ({
  InvitationsListItems: () => <div data-testid="invitations-list" />,
}));

describe("CollaboratorsLink Components", () => {
  const mockProject = { id: 1 };

  beforeEach(() => {
    jest.resetAllMocks();
  });

  describe("CollaboratorsModalBody", () => {
    it("Should show loading state", () => {
      useNestedResource.mockReturnValue({ fetching: true });
      render(<CollaboratorsModalBody project={mockProject} />);
      expect(screen.getByText("Loading collaborators...")).toBeInTheDocument();
    });

    it("Should show error state", () => {
      useNestedResource.mockReturnValue({ fetchable: false });
      render(<CollaboratorsModalBody project={mockProject} />);
      expect(
        screen.getByText("Unable to load collaborators.")
      ).toBeInTheDocument();
    });

    it("Should show content when data is available", () => {
      useNestedResource.mockReturnValue({
        fetchable: true,
        data: {},
        initialised: true,
      });
      render(<CollaboratorsModalBody project={mockProject} />);
      expect(screen.getByTestId("collaborators-list")).toBeInTheDocument();
      expect(screen.getByTestId("invitations-list")).toBeInTheDocument();
    });
  });

  describe("ProjectCollaboratorsLink", () => {
    beforeEach(() => {
      useNestedResource
        .mockReturnValue({
          data: {},
          creating: false,
          fetching: false,
          initialised: true,
          fetchable: true,
        })
        .mockReturnValue({
          data: {},
          creating: false,
          fetching: false,
          initialised: true,
          fetchable: true,
        });
      useProjectPermissions.mockReturnValue({ canEditCollaborators: false });
    });

    it("Should render children and show modal on click", () => {
      render(
        <ProjectCollaboratorsLink project={mockProject}>
          Test Link
        </ProjectCollaboratorsLink>
      );

      fireEvent.click(screen.getByText("Test Link"));
      expect(screen.getByText("Project collaborators")).toBeInTheDocument();
    });

    it("Should disable cancel button during updates", async () => {
      useNestedResource.mockReturnValue({
        data: {
          1: { data: {}, updating: true, deleting: false },
        },
        creating: false,
        fetching: false,
        initialised: true,
        fetchable: true,
      });

      useProjectPermissions.mockReturnValue({
        canEditCollaborators: false,
      });

      render(
        <ProjectCollaboratorsLink project={mockProject}>
          Test Link
        </ProjectCollaboratorsLink>
      );

      fireEvent.click(screen.getByText("Test Link"));

      await waitFor(() => {
        const cancelButton = screen.getByRole("button", { name: /cancel/i });
        expect(cancelButton).toBeDisabled();
      });
    });

    it("Should disable cancel button when collaborator is deleting", async () => {
      useNestedResource.mockReturnValue({
        data: {
          1: { data: {}, updating: false, deleting: true },
        },
        creating: false,
        fetchable: true,
        initialised: true,
      });

      useProjectPermissions.mockReturnValue({
        canEditCollaborators: false,
      });

      render(
        <ProjectCollaboratorsLink project={mockProject}>
          Test Link
        </ProjectCollaboratorsLink>
      );

      fireEvent.click(screen.getByText("Test Link"));

      await waitFor(() => {
        const cancelButton = screen.getByRole("button", { name: /cancel/i });
        expect(cancelButton).toBeDisabled();
      });
    });

    it("Should disable cancel button when any invitation is updating", () => {
      // Mock collaborators (first call)
      useNestedResource.mockReturnValue({
        data: {},
        creating: false,
        fetching: false,
        initialised: true,
        fetchable: true,
      });
      // Mock invitations with an updating invitation (second call)
      useNestedResource.mockReturnValue({
        data: {
          1: { data: {}, updating: true, deleting: false },
        },
        creating: false,
        fetching: false,
        initialised: true,
        fetchable: true,
      });

      render(
        <ProjectCollaboratorsLink project={mockProject}>
          Test Link
        </ProjectCollaboratorsLink>
      );

      // Open the modal
      fireEvent.click(screen.getByText("Test Link"));

      // Check if the cancel button is disabled
      const cancelButton = screen.getByRole("button", { name: /cancel/i });
      expect(cancelButton).toBeDisabled();
    });
    it("Should disable cancel button when any invitation is deleting", () => {
      // Mock collaborators (first call)
      useNestedResource.mockReturnValue({
        data: {},
        creating: false,
        fetching: false,
        initialised: true,
        fetchable: true,
      });
      // Mock invitations with a deleting invitation (second call)
      useNestedResource.mockReturnValue({
        data: {
          1: { data: {}, updating: false, deleting: true },
        },
        creating: false,
        fetching: false,
        initialised: true,
        fetchable: true,
      });

      render(
        <ProjectCollaboratorsLink project={mockProject}>
          Test Link
        </ProjectCollaboratorsLink>
      );

      // Open the modal
      fireEvent.click(screen.getByText("Test Link"));

      // Check if the cancel button is disabled
      const cancelButton = screen.getByRole("button", { name: /cancel/i });
      expect(cancelButton).toBeDisabled();
    });

    it("Should enable cancel button when no updates in progress", () => {
      useNestedResource
        .mockReturnValueOnce({
          data: {},
          creating: false,
          fetchable: true,
          initialised: true,
        })
        .mockReturnValueOnce({
          data: {},
          creating: false,
          fetchable: true,
          initialised: true,
        });

      useProjectPermissions.mockReturnValue({
        canEditCollaborators: false,
      });

      render(
        <ProjectCollaboratorsLink project={mockProject}>
          Test Link
        </ProjectCollaboratorsLink>
      );

      fireEvent.click(screen.getByText("Test Link"));
      const cancelButton = screen.getByRole("button", {
        name: /cancel/i,
      });

      expect(cancelButton).not.toBeDisabled();
    });

    it("Should use large modal for users with edit permissions", () => {
      // Mock permissions to allow editing
      useProjectPermissions.mockReturnValue({ canEditCollaborators: true });

      // Mock resource to avoid loading state
      useNestedResource.mockReturnValue({
        data: {},
        creating: false,
        fetchable: true,
        initialised: true,
      });

      render(
        <ProjectCollaboratorsLink project={mockProject}>
          Test Link
        </ProjectCollaboratorsLink>
      );

      // Open the modal
      fireEvent.click(screen.getByText("Test Link"));

      // Find modal-dialog element and check for modal-lg class
      const modalDialog = document.querySelector(".modal-dialog");
      expect(modalDialog).toHaveClass("modal-lg");
    });

    it("Should close the modal when Cancel button is clicked", async () => {
      // Mock collaborators (first call)
      useNestedResource.mockReturnValueOnce({
        data: {},
        creating: false,
        fetching: false,
        initialised: true,
        fetchable: true,
      });
      // Mock invitations (second call)
      useNestedResource.mockReturnValueOnce({
        data: {},
        creating: false,
        fetching: false,
        initialised: true,
        fetchable: true,
      });

      render(
        <ProjectCollaboratorsLink project={mockProject}>
          Test Link
        </ProjectCollaboratorsLink>
      );

      // Open the modal
      fireEvent.click(screen.getByText("Test Link"));

      // Verify the modal is open
      expect(screen.getByText("Project collaborators")).toBeInTheDocument();

      // Click the Cancel button
      const cancelButton = screen.getByRole("button", { name: /cancel/i });
      fireEvent.click(cancelButton);

      // Wait for the modal to close
      await waitFor(() => {
        expect(
          screen.queryByText("Project collaborators")
        ).not.toBeInTheDocument();
      });
    });
  });
});
