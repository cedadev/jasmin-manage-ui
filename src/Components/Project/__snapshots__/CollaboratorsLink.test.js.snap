// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CollaboratorsLink Components Snapshot tests CollaboratorsModalBody matches the snapshot when data is available 1`] = `
[
  <div
    className="modal-body"
  >
    <div
      className="d-flex justify-content-center my-5"
    >
      <div
        style={
          {
            "fontSize": "110%",
          }
        }
      >
        <i
          className="fas fa-spin fa-sync-alt mr-2 fa-lg"
        />
        Loading collaborators...
      </div>
    </div>
  </div>,
  <div
    className="modal-body"
  >
    <div
      className="fade alert alert-danger show"
      role="alert"
    >
      Unable to load collaborators.
    </div>
  </div>,
  <div
    className="list-group"
    onKeyDown={[Function]}
  >
    <div
      data-testid="collaborators-list"
    />
    <div
      data-testid="invitations-list"
    />
  </div>,
]
`;

exports[`CollaboratorsLink Components Snapshot tests CollaboratorsModalBody matches the snapshot when loading 1`] = `
[
  <div
    className="modal-body"
  >
    <div
      className="d-flex justify-content-center my-5"
    >
      <div
        style={
          {
            "fontSize": "110%",
          }
        }
      >
        <i
          className="fas fa-spin fa-sync-alt mr-2 fa-lg"
        />
        Loading collaborators...
      </div>
    </div>
  </div>,
  <div
    className="modal-body"
  >
    <div
      className="fade alert alert-danger show"
      role="alert"
    >
      Unable to load collaborators.
    </div>
  </div>,
  <div
    className="list-group"
    onKeyDown={[Function]}
  >
    <div
      data-testid="collaborators-list"
    />
    <div
      data-testid="invitations-list"
    />
  </div>,
]
`;

exports[`CollaboratorsLink Components Snapshot tests ProjectCollaboratorsLink matches the snapshot 1`] = `
<button
  className="p-0 btn btn-link"
  disabled={false}
  onClick={[Function]}
  type="button"
>
  <strong>
    Test Link
  </strong>
</button>
`;
